enhancements/microshift/microshift-greenboot.md

```diff
@@ -0,0 +1,321 @@
+be caused by these factors.
+
+If an update requires downloading new container images on a relatively slow
+network, it may take more than the default wait period to run this operation.
+The mitigation would be automatic doubling of the wait period on subsequent
+boots. If this is still not enough, Device Administrators should set the base
+wait period setting to a higher value to start with.
+
+> Container image pull check is performed as a separate step, so there is a
+> clear indication when image download times out due to a slow network.
+
+If any external connectivity is required for upgrades to complete successfully,
+intermittent network outage may result in a rollback to the previous version.
+As a mitigation, Device Administrators should reinstall the upgrade when network
+is available for the Edge Devices.
+
+To further mitigate potential problems related to slow or unstable networks,
+Device Administrators may choose to bundle the upgraded container images in the
+`ostree` software upgrade. In this case, when the upgrade is installed, no
+more container images will need to be downloaded over the network.
+
+Some 3rd party workloads may become operational before the upgrade is declared
+valid and potentially create or update data on the device. If a rollback is
+performed subsequentially, there is a risk of data loss because the file system
+is reverted to its state before the upgrade. One of the ways to mitigate this 
+problem is to have 3rd party workloads wait until a boot is declared successful.
+
+```bash
+$ sudo grub2-editenv - list | grep ^boot_success
+boot_success=1
+```
+
+If any of the user workloads start before the boot is declared successful and
+update host file system or data volumes, there is a risk for the host file system
+and software update to be reverted to its previous state by `greenboot`. In this
+case, user workloads should take measures to ensure their configuration and
+application data consistency. For example, user application may choose to install
+scripts in the `/etc/greenboot/red.d` directory to be invoked by `greenboot` in
+case of an unsuccessful upgrade before the rollback is performed.
+
+> MicroShift implements its own application state and data backup / rollback
+> procedures, but it cannot ensure consistency of 3rd party workloads or data.
+
+### Drawbacks
+
+The technical implementation of upgrade validation is rather simple as it relies
+on the existing `greenboot` and `rpm-ostree` facilities. However, end users may
+find it annoying to have their Edge Devices rebooted few times until the upgrade
+is either successful or rolled back.
+
+It is possible that certain 3rd party workloads become operational before the
+upgrade is declared successful. This situation may further hurt user experience
+if their Edge Devices are rebooted due to MicroShift core functionality
+validation not passing.
+
+## Design Details
+
+### Open Questions
+N/A
+
+### Test Plan
+N/A
```

Comment:
What sort of automated testing can we describe?

Reply:
Added description of possible tests, also part of dev-docs in the implementation.